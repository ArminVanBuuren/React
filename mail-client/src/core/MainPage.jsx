
import React from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';

import combinedReducer from '../redux/reducers.js';
import PagesRouter from './PagesRouter.jsx';

import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
let store = createStore(combinedReducer, applyMiddleware(thunk));

class MainPage extends React.PureComponent {

  static propTypes = {
    emulation: PropTypes.bool,
  };

  render() {
    return (
      <Provider store={store}>
        <PagesRouter emulation={this.props.emulation} />
      </Provider>
    );
  }
}

export default MainPage;
