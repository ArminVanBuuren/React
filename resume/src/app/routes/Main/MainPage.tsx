import React, { Fragment, useEffect } from 'react';
import { Helmet } from 'react-helmet';
import { useLocalize } from '@hooks/useLocalize';
import Loading from '@components/Loading/Loading';
import { useStores } from '@stores/store';
import useStyles from './MainPage.jss';
import LinkButton from '@components/LinkButton/LinkButton';
import Guthub from '@icons/Github';
import Skype from '@icons/Skype';
import Telegram from '@icons/Telegram';
import { getPhoneNumber } from '@utils/formatter';
import { toJS } from 'mobx';
import { useTranslation } from 'react-i18next';
import SimpleCard from '@components/Cards/SimpleCards';
import classNames from 'classnames';
import PhoneIphone from '@material-ui/icons/PhoneIphone';
import Linkedin from '@icons/Linkedin';

const MainPage = () : JSX.Element => {
	const classes = useStyles();
	const { i18n } = useTranslation();
	const { commonStore, cvStore } = useStores();
	const { cvData, getCvData } = cvStore;

	useEffect(() => {
		getCvData().catch(e => console.log(e));
	}, []);


	return (
		<Loading isReady={Boolean(cvData)} classes={classes.root}>
			<Helmet>
				<title>{ process.env.APP_NAME }</title>
			</Helmet>
			{ Object.entries(cvData ?? {}).map(([name, cvData]) => {
				const data = cvData.localized.find(x => x.langMatches.includes(i18n.language));
				if (!data)
					return false;

				return (
					<SimpleCard key={name} classes={classes.card}>
						<header className={classes.header}>
							<div className={classes.headerImg}>
								<img className={classNames('elevation-z10')} src={cvData.avatar} />
							</div>
							<div className={classes.headerInfo}>
								<h1 className='font-bold'>{data.name}</h1>
								<h2 className='font-normal'>{data.position}</h2>
								<ul>
									{ data.contacts.tel?.map(x => {
										const { normal, display } = getPhoneNumber(x);
										return <li key={normal}><LinkButton to={{pathname:`tel:${normal}`}} target="_blank"><PhoneIphone className={classes.itemImg}/>{display}</LinkButton></li>;
									}) }
									{ data.contacts.skype?.map(x => <li key={x}><LinkButton to={{pathname:`skype:${x}?userinfo`}} target="_blank"><Skype classes={classes.itemImg}/>{x}</LinkButton></li>) }
									{ data.contacts.github?.map(x => <li key={x}><LinkButton to={{pathname:`https://github.com/${x}`}} target="_blank"><Guthub classes={classes.itemImg}/>{x}</LinkButton></li>) }
									{ data.contacts.telegram?.map(x => <li key={x}><LinkButton to={{pathname:`https://t.me/${x}`}} target="_blank"><Telegram classes={classes.itemImg}/>{x}</LinkButton></li>) }
									{ data.contacts.linkedin?.map(x => <li key={x}><LinkButton to={{pathname:`https://t.me/${x}`}} target="_blank"><Linkedin classes={classes.itemImg}/>{x}</LinkButton></li>) }
								</ul>
							</div>
						</header>
						<main>

						</main>
						<div className={classes.title}>

							{/*<hr/>*/}
						</div>
					</SimpleCard>
				);
			}) }
		</Loading>
	);
};

export default MainPage;
