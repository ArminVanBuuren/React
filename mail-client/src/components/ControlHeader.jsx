import React, {useState} from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { pure } from 'recompose';

import CachedIcon from '@material-ui/icons/Cached';
import AppBar from "@material-ui/core/AppBar";
import Toolbar from "@material-ui/core/Toolbar";
import IconButton from "@material-ui/core/IconButton";
import Typography from "@material-ui/core/Typography";
import InputBase from "@material-ui/core/InputBase";
import { fade, makeStyles } from "@material-ui/core/styles";
import SearchIcon from "@material-ui/icons/Search";
import EmailIcon from '@material-ui/icons/Email';

import { mailItemsFetchAC } from '../redux/fetchThunk';

import { ACTION_TYPES, ACTION_MODE, searchingTextAct, createMessageAct } from '../redux/countersAC';

const useStyles = makeStyles((theme) => ({
  root: {
    flexGrow: 1
  },
  menuButton: {
    marginRight: theme.spacing(2)
  },
  title: {
    flexGrow: 1,
    display: "none",
    [theme.breakpoints.up("sm")]: {
      display: "block"
    }
  },
  search: {
    position: "relative",
    borderRadius: theme.shape.borderRadius,
    backgroundColor: fade(theme.palette.common.white, 0.15),
    "&:hover": {
      backgroundColor: fade(theme.palette.common.white, 0.25)
    },
    marginLeft: 0,
    width: "100%",
    [theme.breakpoints.up("sm")]: {
      marginLeft: theme.spacing(1),
      width: "auto"
    }
  },
  searchIcon: {
    padding: theme.spacing(0, 2),
    height: "100%",
    position: "absolute",
    pointerEvents: "none",
    display: "flex",
    alignItems: "center",
    justifyContent: "center"
  },
  inputRoot: {
    color: "inherit"
  },
  inputInput: {
    padding: theme.spacing(1, 1, 1, 0),
    // vertical padding + font size from searchIcon
    paddingLeft: `calc(1em + ${theme.spacing(4)}px)`,
    transition: theme.transitions.create("width"),
    width: "100%",
    [theme.breakpoints.up("sm")]: {
      width: "12ch",
      "&:focus": {
        width: "20ch"
      }
    }
  },
  sectionDesktop: {
    display: "none",
    [theme.breakpoints.up("sm")]: {
      display: "flex"
    }
  },
}));

function IntControlHeader(props) {
  const { emulation, boxName, dispatch, searchingText } = props;
  const classes = useStyles();

  return (
    <div className={classes.root}>
      <AppBar position="static">
        <Toolbar>
          <Typography className={classes.title} variant="h6" noWrap>{boxName}</Typography>
          <IconButton id="createButton" color="inherit" disabled={!boxName} onClick={(EO) => { dispatch( createMessageAct({}) ); }} >
            <EmailIcon />
          </IconButton>
          <IconButton id="reloadButton" color="inherit" onClick={(EO) => { dispatch( mailItemsFetchAC(dispatch, emulation) ); }}>
            <CachedIcon />
          </IconButton>
          <div className={classes.search}>
            <div className={classes.searchIcon}>
              <SearchIcon color="inherit" />
            </div>
            <InputBase
              id="searchText"
              placeholder="Search…"
              classes={{root: classes.inputRoot, input: classes.inputInput, }}
              inputProps={{ "aria-label": "search" }}
              value={searchingText == undefined ? "" : searchingText}
              onChange={(EO) => { dispatch(searchingTextAct(EO.target.value))  }}
            />
          </div>
        </Toolbar>
      </AppBar>
    </div>
  );
}

IntControlHeader.propTypes = {
  emulation: PropTypes.bool,
  boxName: PropTypes.string.isRequired,
  searchingText: PropTypes.string,
};

const mapStateToProps = function (state) {
    return {
      // из раздела Redux с именем counter свойство cnt будет доступно
      // данному компоненту как this.props.cnt
      emulation: state.counters.emulation,
      boxName: state.counters.boxName,
      searchingText: state.counters.searchingText,
    };
  };

// присоединяем (connect) компонент к хранилищу Redux
const ControlHeader = connect(mapStateToProps)(pure(IntControlHeader));

export default ControlHeader;
