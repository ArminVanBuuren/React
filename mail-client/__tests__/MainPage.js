"use strict";

import React from 'react';
import * as renderer from 'react-test-renderer';
import { URLSearchParams } from 'url';

import MainPage from '../src/core/MainPage.jsx';
import ControlHeader from '../src/components/ControlHeader.jsx';

global.URLSearchParams = URLSearchParams;
global.requestAnimationFrame = function(callback) {
  setTimeout(callback, 0);
};


// console.log(`Start:[${formatDateTime(new Date())}]`);
// it('mock setTimeout test', (done) => {
//   setTimeout(() => {
//     done();
//     console.log(`End:[${formatDateTime(new Date())}]`);
//   }, 1000);
// });

// it('mock setTimeout test', (done) => {
//   setTimeout(() => {
//     done();
//     console.log(`End:[${formatDateTime(new Date())}]`);
//   }, 1000);
// });



function clickButton(component, buttonId){
  const buttonElem = component.root.findAll( btn => btn.id === buttonId )[0];
  console.log(`${buttonId} finded=${buttonElem != null && buttonElem != undefined}`);
  buttonElem.props.onClick();
}

function formatDateTime(dt) {
  var year = dt.getFullYear();
  var month = dt.getMonth()+1;
  var day = dt.getDate();
  var hours = dt.getHours();
  var minutes = dt.getMinutes();
  var seconds = dt.getSeconds();
  return str0l(day, 2) + '.' + str0l(month, 2) + '.' + year + ' ' + str0l(hours, 2) + ':' + str0l(minutes, 2) + ':' + str0l(seconds, 2);
}

function str0l(val, len) {
  var strVal=val.toString();
  while ( strVal.length < len )
      strVal='0'+strVal;
  return strVal;
}

test('работа MainPage', async () => {
  let component = renderer.create( <MainPage emulation={true} />);
  console.log(`Start:[${formatDateTime(new Date())}]`);
  await new Promise((r) => setTimeout(r, 2000));
  console.log(`End:[${formatDateTime(new Date())}]`);
  let componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();

  // clickButton(component, "createButton");
  // await new Promise((r) => setTimeout(r, 1000));
  // let componentCreate = component.toJSON();
  // expect(componentCreate).toMatchSnapshot();
});



function whenPromise(description, delay, func) {
  return new Promise(async (resolve, reject) => {
    try {
      console.log(`Start:[${formatDateTime(new Date())}]=${description}`);
      let result = func();

      jest.useRealTimers();
      setTimeout(() => {
        console.log(`End:[${formatDateTime(new Date())}]`);
        resolve(result);
      }, delay);
      jest.runAllTimers();

      //await new Promise((r) => setTimeout(r, delay));
      // console.log(`End:[${formatDateTime(new Date())}]`);
      // resolve(result);
    }
    catch (error) {
      console.error(`error=${error}`);
      reject(error);
    }
  });
}

test('работа MainPage', () => {
  return;

  whenPromise("Получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась", 5000, function () { 
    return renderer.create( <MainPage emulation={true} /> );
  })
  .then((component) => {
    let componentTree = component.toJSON();
    expect(componentTree).toMatchSnapshot();

    return whenPromise("Нажимаем на кнопку создать", 2000, () => {
      clickButton(component, "createButton");
      return component;
    });
  })
  .then((component) => {
    let componentCreate = component.toJSON();
    expect(componentCreate).toMatchSnapshot(); 

    return whenPromise("Нажимаем на кнопку обновить", 1000, () => {
      clickButton("reloadButton");
      return component;
    });
  })
  .then((component) => {
    let componentTree = component.toJSON();
    expect(componentTree).toMatchSnapshot();
    return null;
  })
  .catch((error) =>{
    console.log(error);
  });

  
  /*
  // можно эмулировать события, передавая в качестве объекта события то что нам нужно:
  wrapper.find('select').simulate('change', {
    target: { value: "hello" },
  });
  */

});

