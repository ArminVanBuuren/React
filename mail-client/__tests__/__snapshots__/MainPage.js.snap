// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа MainPage 1`] = `
<div
  className="root"
>
  <div
    id="waiting"
    style={
      Object {
        "opacity": 0,
        "zIndex": -1,
      }
    }
  >
    <div
      className="loader-container"
    >
      <div
        className="loader-div"
      >
        <div
          className="loader-div"
        >
          <div
            className="loader-div"
          >
            <div
              className="loader-div"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
